rust/arrow/src/json/writer.rs

```diff
@@ -0,0 +1,301 @@
+            }
+        }
+        DataType::Struct(_) => {
+            let arr = as_struct_array(array);
+            let inner_col_names = arr.column_names();
+
+            let mut inner_objs = iter::repeat(JsonMap::new())
+                .take(row_count)
+                .collect::<Vec<JsonMap<String, Value>>>();
+
+            arr.columns()
+                .iter()
+                .enumerate()
+                .for_each(|(j, struct_col)| {
+                    set_column_for_json_rows(
+                        &mut inner_objs,
+                        row_count,
+                        struct_col,
+                        inner_col_names[j],
+                    );
+                });
+
+            rows.iter_mut()
+                .take(row_count)
+                .zip(inner_objs.into_iter())
+                .for_each(|(row, obj)| {
+                    row.insert(col_name.to_string(), Value::Object(obj));
+                });
+        }
+        _ => {
+            panic!(format!("Unsupported datatype: {:#?}", array.data_type()));
+        }
+    }
+}
+
+pub fn record_batches_to_json_rows(
+    batches: &[RecordBatch],
+) -> Vec<JsonMap<String, Value>> {
+    let mut rows: Vec<JsonMap<String, Value>> = iter::repeat(JsonMap::new())
+        .take(batches.iter().map(|b| b.num_rows()).sum())
+        .collect();
+
+    if !rows.is_empty() {
+        let schema = batches[0].schema();
+        let mut base = 0;
+        batches.iter().for_each(|batch| {
+            let row_count = batch.num_rows();
+            batch.columns().iter().enumerate().for_each(|(j, col)| {
+                let col_name = schema.field(j).name();
+                set_column_for_json_rows(&mut rows[base..], row_count, col, col_name);
+            });
+            base += row_count;
+        });
+    }
+
+    rows
+}
+
+/// A JSON writer
+#[derive(Debug)]
+pub struct Writer<W: Write> {
+    writer: BufWriter<W>,
+}
+
+impl<W: Write> Writer<W> {
+    pub fn new(writer: W) -> Self {
+        Self::from_buf_writer(BufWriter::new(writer))
+    }
+
+    pub fn from_buf_writer(writer: BufWriter<W>) -> Self {
+        Self { writer }
+    }
+
+    pub fn write_row(&mut self, row: &Value) -> Result<()> {
+        self.writer.write_all(&serde_json::to_vec(row)?)?;
+        self.writer.write_all(b"\n")?;
+        Ok(())
+    }
+
+    pub fn write_batches(&mut self, batches: &[RecordBatch]) -> Result<()> {
+        for row in record_batches_to_json_rows(batches) {
+            self.write_row(&Value::Object(row))?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs::{read_to_string, File};
+    use std::sync::Arc;
+
+    use crate::json::reader::*;
+
+    use super::*;
+
+    #[test]
+    fn write_simple_rows() {
+        let schema = Schema::new(vec![
+            Field::new("a", DataType::Int32, false),
+            Field::new("b", DataType::Utf8, false),
+        ]);
+
+        let a = Int32Array::from(vec![1, 2, 3, 4, 5]);
```

Comment:
could have use `vec![Some(1), None, Some(3), Some(4,) Some(5)]`, so that we also test null values?